Source/WebCore/html/HTMLElement.cpp

```diff
@@ -1234,19 +1234,43 @@ static ExceptionOr<void> checkPopoverValidity(Element& element, PopoverVisibilit
     return { };
 }
 
-static HTMLElement* topmostPopoverAncestor(Element& element)
-{
-    for (auto& element : lineageOfType<HTMLElement>(element)) {
-        if (element.popoverState() != PopoverState::Auto)
-            continue;
-        if (element.popoverData()->visibilityState() != PopoverVisibilityState::Showing)
+// https://html.spec.whatwg.org/#topmost-popover-ancestor
+// Consider both DOM ancestors and popovers where the given popover was invoked from as ancestors.
+// Use top layer positions to disambiguate the topmost one when both exist.
+static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+{
+    // Store indices to avoid having to do O(n) search for every popover invoker.
+    HashMap<Element*, size_t> topLayerPositions;
```

Comment:
While this instance appears safe, I think we discourage the use of raw pointers in this manner.

cc @rniwa

Reply:
Yeah, definitely don't use raw pointers. Use RefPtr<Element> or Ref<Element>.